File: Stack/PostfixtoPrefix.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;


string postfixToPrefix(string &s){
    // Write your code here.
    string res = "";
    stack<string> st;
    for(int i = 0 ;i < s.length(); i++){
        if(isalpha(s[i]) || isdigit(s[i])){
            string temp(1,s[i]);
            st.push(temp);
        }

        else{
            string op1 = st.top();
            st.pop();
            string op2 = st.top();
            st.pop();
            string temp = s[i] + op2 + op1;
            st.push(temp);
        }
    }
    res = st.top();
    return res;
}




JAVA - 
    
class Solution {
    static String postToPre(String post_exp) {
        // code here
        int n = post_exp.length();
        Stack<String> st = new Stack<>();
        
        for(int i = 0 ; i < n ; i++){
            char ch = post_exp.charAt(i);
            if(Character.isLetter(ch)){
                st.push(String.valueOf(ch));
            }else{
                String c1 = st.pop();
                String c2 = st.pop();
                String temp = ch + c2 + c1;
                st.push(temp);
            }
        }
        return st.pop();
    }
}

File: Stack/Infix,Prefix,Postfix/PostfixtoPrefix.java
class Solution {
    // Function to convert a postfix expression to a prefix expression.
    static String postToPre(String post_exp) {
        int n = post_exp.length(); // Get the length of the postfix expression
        Stack<String> st = new Stack<>(); // Initialize a stack to hold intermediate results
        
        // Loop through each character in the postfix expression
        for (int i = 0; i < n; i++) {
            char ch = post_exp.charAt(i); // Get the current character
            
            // If the character is a letter, push it onto the stack
            if (Character.isLetter(ch)) {
                st.push(String.valueOf(ch));
            } 
            // If the character is an operator
            else {
                // Pop two elements from the stack
                String c1 = st.pop();
                String c2 = st.pop();
                
                // Form a new string with the operator followed by the two popped elements
                String temp = ch + c2 + c1;
                
                // Push the new string back onto the stack
                st.push(temp);
            }
        }
        // The final result is the only element left in the stack
        return st.pop();
    }
}
